File: practices/chapter16_dynamic_plan/python/ch_16_calculate_subset.py
def generate_subset(dataList=[]):
  final_subset = [[]]
  for data in dataList:
    final_subset.extend([subset + [item] for subset in final_subset])
  return final_subset

def get_best_choice(dataList=[], max_weight=4):
  subset_items = generate_subset(dataList)
  max_val = 0
  final_items_choice = []
  for items in subset_items:
    if items:
      weight_sum = 0
      value_sum = 0
      # sum items
      for item in items:
        weight_sum += item['weight']
        value_sum += item['value']
        if weight_sum <= max_weight:
          if value_sum > max_val:
            max_val = value_sum
            final_items_choice = items
  
  return max_val, final_items_choice

def main():
  items = [
    {
      'name': 'TV',
      'value': 40000,
      'weight': 3,
    },
    {
      'name': 'SPEAKER',
      'value': 50000,
      'weight': 4,
    },
    {
      'name': 'LAPTOP',
      'value': 20000,
      'weight': 1,
    }
  ]

  max_val, final_items_choice = get_best_choice(items, 4)
  print('Items = {}, \nValues = {}'.format(final_items_choice, max_val))

if __name__ == "__main__":
    main()

File: practices/chapter16_dynamic_plan/js/ch_16_calculate_subset.js
function generateSubset(dataList=[]) {
  let res = [[]]

  dataList.forEach(data => {
    const latest = res.map(r => [...r, data])
    res.push(...latest)
  });

  return res
}

function getBestChoice(dataList=[], maxWeight=4) {
  const subsetItems = generateSubset(dataList)
  let maxVal = 0
  let finalItems = []

  for (const items of subsetItems) {
    if(items.length > 0) {
      let weightSum = 0
      let valueSum = 0
      for (const item of items) {
        weightSum += item.weight
        valueSum += item.value
      }
      
      if(weightSum <= maxWeight) {
        if(valueSum > maxVal) {
          maxVal = valueSum
          finalItems = items
        }
      }
    }
  }

  return ({
    maxVal,
    finalItems,
  })
}

function main() {
  const items = [
    {
      'name': 'TV',
      'value': 40000,
      'weight': 3,
    },
    {
      'name': 'SPEAKER',
      'value': 50000,
      'weight': 4,
    },
    {
      'name': 'LAPTOP',
      'value': 20000,
      'weight': 1,
    }
  ]

  const {
    maxVal, finalItems,
  } = getBestChoice(items, 4)
  console.log(`Items: ${JSON.stringify(finalItems)}, \nValues: ${maxVal}`)
}

main()